ggu/main.c

```diff
@@ -16,11 +16,313 @@
+		rte_memcpy(&policy->flow.f.v6, policy_ptr,
+				sizeof(policy->flow.f.v6));
+		policy_ptr = policy_ptr + sizeof(policy->flow.f.v6);
+	} else {
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"ggu: parse policy with unknown network layer protocol %hhu\n",
+			policy->flow.proto);
+		return -1;
+	}
+
+	rte_memcpy(&policy->params, policy_ptr, sizeof(policy->params));
+	policy_ptr = policy_ptr + sizeof(policy->params);
+
+	return 0;
+}
+
+static void
+process_single_policy(struct ggu_policy *policy, struct ggu_config *ggu_conf)
+{
+	/*
+	 * Calculate the RSS hash value for the
+	 * pair <Src, Dst> in the decision.
+	 */
+	uint32_t rss_hash_val = rss_ip_flow_hf(&policy->flow, 0, 0);
+
+	/*
+	 * Identify which GK block is responsible for the
+	 * pair <Src, Dst> in the decision.
+	 */
+	int block_id = -1;
+	uint32_t i;
+	uint32_t idx;
+	uint32_t shift;
+	uint16_t queue_id;
+	struct gk_config *gk_conf = get_gk_conf();
+	struct gk_cmd_entry entry;
+
+	/*
+	 * XXX Change the mapping rss hash value to rss reta entry
+	 * if the reta size is not 128.
+	 */
+	RTE_ASSERT(ggu_conf->rss_conf.reta_size == 128);
+	rss_hash_val = (rss_hash_val & 127);
+
+	idx = rss_hash_val / RTE_RETA_GROUP_SIZE;
+	shift = rss_hash_val % RTE_RETA_GROUP_SIZE;
+	queue_id = ggu_conf->rss_conf.reta_conf[idx].reta[shift];
+
+	/* XXX Change mapping queue id to the gk instance id efficiently. */
+	for (i = gk_conf->lcore_start_id; i<= gk_conf->lcore_end_id; i++)
+		if (gk_conf->instances[i - gk_conf->lcore_start_id].
+				rx_queue_front == queue_id)
+			block_id = i - gk_conf->lcore_start_id;
+
+	if (block_id == -1) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"ggu: wrong RSS configuration for GK blocks!\n");
+		return;
+	}
+
+	/*
+	 * Obtain mailbox of that GK block,
+	 * and send the policy decision to the GK block.
+	 */
+	entry.op = GGU_POLICY_ADD;
+	rte_memcpy(&entry.u.ggu.flow, &policy->flow, sizeof(policy->flow));
+	entry.u.ggu.params.state = policy->params.state;
+	entry.u.ggu.params.tx_rate_kb_cycle =
+		rte_be_to_cpu_32(policy->params.tx_rate_kb_cycle);
+	entry.u.ggu.params.grantor_id =
+		rte_be_to_cpu_32(policy->params.grantor_id);
+	entry.u.ggu.params.cap_expire_at =
+		rte_be_to_cpu_64(policy->params.cap_expire_at);
+	entry.u.ggu.params.renewal_step_cycle =
+		rte_be_to_cpu_64(policy->params.renewal_step_cycle);
+
+	send_gk_cmd(block_id, &entry, gk_conf);
+}
+
+static void
+process_single_packet(struct rte_mbuf *pkt, struct ggu_config *ggu_conf)
+{
+	uint8_t j;
+	uint8_t *data;
+	uint8_t *policy_ptr;
+	uint8_t num_policies;
+	uint16_t *proto_ptr;
+	uint16_t ether_type;
+	struct ether_hdr *eth_hdr;
+	struct ipv4_hdr *ip4hdr;
+	struct ipv6_hdr *ip6hdr;
+	struct udp_hdr *udphdr;
+
+	eth_hdr = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	ether_type = rte_be_to_cpu_16(eth_hdr->ether_type);
+
+	switch (ether_type) {
+	case ETHER_TYPE_IPv4:
+		ip4hdr = rte_pktmbuf_mtod_offset(pkt, 
+					struct ipv4_hdr *, 
+					sizeof(struct ether_hdr));
+		if (ip4hdr->next_proto_id != IPPROTO_UDP)
```

Comment:
Sorry, I just remembered that the protocol field is only a byte, so no problem here.